### PR TITLE
Add rewards owed

### PR DESCRIPTION
**Pull request by smbsp**
Changes:

1. Added a function getStakerCurrentReward() for FrontEnd to view the staker's current accumulated reward.
2. Added a require check to make sure the stop functionality isn't called multiple times.
3. Modified < to <= for current timestamp check. It can hardly be a possibility that the collectRewards() is triggered exactly (to the second) on the 14th day. Now with the modification, even if it does, we are covered.
4. Modified the test cases as per the changes

closes: #347

### CODE REVIEW

File: contracts/governance/StakingRewards/StakingRewards.sol [hunk 1]
```diff
@@ -82,7 +83,7 @@ contract StakingRewards is StakingRewardsStorage, Initializable {
 			require(currentTS > withdrawals[staker], "allowed after 14 days");
 		}
 
-		for (uint256 i = withdrawals[staker]; i < currentTS && i < withdrawals[staker].add(maxDuration); i += TWO_WEEKS) {
+		for (uint256 i = withdrawals[staker]; i <= currentTS && i < withdrawals[staker].add(maxDuration); i += TWO_WEEKS) {
```
Comment (tjcloa) [inline contracts/governance/StakingRewards/StakingRewards.sol:86]: the period is considered passed the next second after the 2-weeks multiple reached, not on the same second
```suggestion
		for (uint256 i = withdrawals[staker]; i < currentTS && i < withdrawals[staker].add(maxDuration); i += TWO_WEEKS) {
```

File: contracts/governance/StakingRewards/StakingRewards.sol [hunk 2]
```diff
@@ -149,4 +151,22 @@ contract StakingRewards is StakingRewardsStorage, Initializable {
...
+
+	/**
+	 * @notice Get staker's current accumulated reward
+	 * @dev Collecting rewards can yeild slightly different values as there are mandatory checks
+	 * while withdrawing:
+	 * 1. User cannot withdraw within 14 days from the start/last withdrawn date
+	 * 2. The rewards can only be processed for a max duration at a time
+	 * @return The accumulated reward
+	 */
+	function getStakerCurrentReward() external view returns (uint256 amount) {
```
Comment (tjcloa) [inline contracts/governance/StakingRewards/StakingRewards.sol:163]: pls use this function in collectRewards() else it is code duplication

File: contracts/governance/StakingRewards/StakingRewards.sol [hunk 3]
```diff
@@ -149,4 +136,43 @@ contract StakingRewards is StakingRewardsStorage, Initializable {
...
+		} else {
+			duration = currentTS;
+		}
+
+		for (uint256 i = withdrawalTime; i <= currentTS && i <= withdrawalTime.add(duration); i += TWO_WEEKS) {
+			count++;
+			weightedStake = weightedStake.add(_computeRewardForDate(staker, lastFinalisedBlock, i));
+		}
+
+		require(weightedStake > 0, "weightedStake is zero");
```
Comment (tjcloa) [inline contracts/governance/StakingRewards/StakingRewards.sol:174]: why not just return 0?
```suggestion
```

File: contracts/governance/StakingRewards/StakingRewards.sol [hunk 4]
```diff
@@ -149,4 +136,43 @@ contract StakingRewards is StakingRewardsStorage, Initializable {
...
+
+	/**
+	 * @notice Get staker's current accumulated reward
+	 * @dev The collectReward() function internally calls this function to calculate reward amount
+	 * @param isMaxDuration True: Runs for the maximum duration
+	 * False: Runs till the current timestamp
+	 * @return The timestamp of last withdrawal
+	 * @return The accumulated reward
+	 */
+	function getStakerCurrentReward(bool isMaxDuration) public view returns (uint256 withdrawalTime, uint256 amount) {
```
Comment (tjcloa) [inline contracts/governance/StakingRewards/StakingRewards.sol:148]:
```suggestion
	function getStakerCurrentReward(bool considerMaxDuration) public view returns (uint256 withdrawalTime, uint256 amount) {
```

File: contracts/governance/StakingRewards/StakingRewards.sol [hunk 5]
```diff
@@ -149,4 +136,43 @@ contract StakingRewards is StakingRewardsStorage, Initializable {
...
+		uint256 duration;
+		if (withdrawals[staker] == 0) {
+			require((currentTS.sub(startTime)) > TWO_WEEKS, "allowed after 14 days of start");
+			withdrawalTime = startTime;
+		} else {
+			require(currentTS > withdrawals[staker], "allowed after 14 days");
+			withdrawalTime = withdrawals[staker];
+		}
+
+		if (isMaxDuration) {
```
Comment (tjcloa) [inline contracts/governance/StakingRewards/StakingRewards.sol:163]:
```suggestion
		if (considerMaxDuration) {
```

File: contracts/governance/StakingRewards/StakingRewards.sol [hunk 6]
```diff
@@ -149,4 +136,43 @@ contract StakingRewards is StakingRewardsStorage, Initializable {
 		require(SOV.transfer(_receiver, _amount), "transfer failed");
 		emit RewardWithdrawn(_receiver, _amount);
 	}
+
+	/**
+	 * @notice Get staker's current accumulated reward
+	 * @dev The collectReward() function internally calls this function to calculate reward amount
+	 * @param isMaxDuration True: Runs for the maximum duration
```
Comment (tjcloa) [inline contracts/governance/StakingRewards/StakingRewards.sol:143]:
```suggestion
	 * @param considerMaxDuration True: Runs for the maximum duration - used in tx not to run out of gas, false - to query total rewards
```

File: contracts/governance/StakingRewards/StakingRewards.sol (hunk 2 above)
Comment (smbsp) [reply contracts/governance/StakingRewards/StakingRewards.sol]: Done

File: contracts/governance/StakingRewards/StakingRewards.sol (hunk 3 above)
Comment (smbsp) [reply contracts/governance/StakingRewards/StakingRewards.sol]: Done

File: contracts/governance/StakingRewards/StakingRewards.sol (hunk 4 above)
Comment (smbsp) [reply contracts/governance/StakingRewards/StakingRewards.sol]: Done

File: contracts/governance/StakingRewards/StakingRewards.sol (hunk 5 above)
Comment (smbsp) [reply contracts/governance/StakingRewards/StakingRewards.sol]: Done

File: contracts/governance/StakingRewards/StakingRewards.sol (hunk 6 above)
Comment (smbsp) [reply contracts/governance/StakingRewards/StakingRewards.sol]: Done

**Review by tjcloa — APPROVED** [review]

**Pull request merged**